src/mpyl/run_plan.py

```diff
@@ -161,16 +161,57 @@ def write_to_pickle_file(self):
             logger.info(f"Storing run plan in: {RUN_PLAN_PICKLE_FILE}")
             pickle.dump(self, file, pickle.HIGHEST_PROTOCOL)
 
+    def write_to_summary_file(self):
+        summary = "| 👷 Project | 🏗 Build | 🧪 Test | 🚀 Deploy | 🦺 Post-deploy |\n"
+        summary += "| ---------- | :------: | :-----: | :-------: | :------------: |\n"
+
+        def get_icon(project_execution: ProjectExecution, stage: str):
+            if project_execution.project.pipeline == "docker":
+                return "🐳"
+            if (
+                project_execution.project.pipeline == "sbt"
+                and project_execution
+                in self.get_executions_for_stage_name(stage, use_full_plan=True)
+            ):
+                return "💾" if project_execution.cached else "☕️"
+            return ""
+
+        all_executions = self._get_all_executions(use_full_plan=True)
+        if len(all_executions) == 0:
+            summary = "Nothing to do 🤷\n"
+
+        for execution in all_executions:
+            build_plan = get_icon(execution, "build")
+            test_plan = get_icon(execution, "test")
+            deploy_plan = (
+                "🚀"
+                if execution
+                in self.get_executions_for_stage_name("deploy", use_full_plan=True)
+                else ""
+            )
+            postdeploy_plan = (
+                "🦺"
+                if execution
+                in self.get_executions_for_stage_name("postdeploy", use_full_plan=True)
+                else ""
+            )
+            summary += f"| {execution.name} | {build_plan} | {test_plan} | {deploy_plan} | {postdeploy_plan} |\n"
+
+        logger = logging.getLogger("mpyl")
+        os.makedirs(os.path.dirname(RUN_PLAN_SUMMARY_FILE), exist_ok=True)
+        with open(RUN_PLAN_SUMMARY_FILE, "w", encoding="utf-8") as file:
+            logger.info(f"Storing run plan summary in: {RUN_PLAN_SUMMARY_FILE}")
+            file.write(summary)
+
     def write_to_json_file(self):
         run_plan: dict = {}
 
-        for stage, executions in self._full_plan.items():
+        for executions in self._selected_plan.values():
             for execution in executions:
-                stages: list[dict[str, Union[str, bool]]] = run_plan.get(
-                    execution.project.name, {}
-                ).get("stages", [])
-                stages.append({"name": stage.name, "cached": execution.cached})
-
+                stages = {
+                    stage.name: not execution.cached  # 'should it run this stage' value
```

Comment:
I don't have an easy way to showcase it, but I don't see how this code works as expected.

Say we have a plan that looks like this:
```python
{
  "build": [
    ProjectExecution("project_a", cached=True),
    ProjectExecution("project_b", cached=False),
  ],
  "test": [
    ProjectExecution("project_a", cached=False),
  ]
}
```

Then we expect this run plan JSON to be (simplified):
```json
{
  "project_a": {
    "build": false,
    "test": true
   },
   "project_b": {
     "build": true,
   }
}
```

---

However, if I'm reading this code right, then:
```python
for executions in self._selected_plan.values():
# self._selected_plan.values() = [ 
#    ProjectExecution("project_a", cached=True),
#    ProjectExecution("project_b", cached=False),
#    ProjectExecution("project_a", cached=False),
# ]
  for execution in executions:

  # ------ first iteration ----
  # execution = ProjectExecution("project_a", cached=True),
    stages = {
      stage.name: not execution.cached  # 'should it run this stage' value
      for stage in self._full_plan.keys()
    }
    # stages = {
    #   "build": False
    #   "test": False
    # }
    
    run_plan.update(… | stages )
    # run_plan = {
    #   "project_a": {
    #     "build: False,
    #     "test": False
    #   }
    # } 
 
  # ------ second iteration ----
  # execution = ProjectExecution("project_b", cached=False),
    stages = {
      stage.name: not execution.cached  # 'should it run this stage' value
      for stage in self._full_plan.keys()
    }
    # stages = {
    #   "build": True
    #   "test": True
    # }
    
    run_plan.update(… | stages )
    # run_plan = {
    #   "project_a": {
    #     "build: False,
    #     "test": False
    #   },
    #   "project_b": {
    #     "build": True,
    #     "test": True
    #   }
    # }

  # ------ third iteration ----
  # execution = ProjectExecution("project_a", cached=False),
    stages = {
      stage.name: not execution.cached  # 'should it run this stage' value
      for stage in self._full_plan.keys()
    }
    # stages = {
    #   "build": True
    #   "test": True
    # }
    
    run_plan.update(… | stages )
    # run_plan = {
    #   "project_a": {      <--- the previous value is overriden
    #     "build: True,
    #     "test": True
    #   },
    #   "project_b": {
    #     "build": True,
    #     "test": True
    #   }
    # }
```

Which is completely different than the expected JSON.

Reply:
In either case, this would really benefit from testing.

Reply:
You're very right!